<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Gene dropping simulations - haplotypes — gen.simuHaplo • GENLIB</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Gene dropping simulations - haplotypes — gen.simuHaplo" />
<meta property="og:description" content="Gene dropping from founders to specified probands. If sequence data is provided the function can translate results to sequences, otherwise will produce proband haplotypes with segments labelled w.r.t. the founder the segment came from" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GENLIB</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Gene dropping simulations - haplotypes</h1>
    
    <div class="hidden name"><code>gen.simuHaplo.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Gene dropping from founders to specified probands. If sequence data is provided the function can translate results to sequences, otherwise will produce proband haplotypes with segments labelled w.r.t. the founder the segment came from</p>
    </div>

    <pre class="usage"><span class='fu'>gen.simuHaplo</span><span class='op'>(</span><span class='va'>gen</span>, <span class='va'>pro</span>, <span class='va'>ancestors</span>, simulNo <span class='op'>=</span> <span class='fl'>1</span>, RecombRate<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>0</span><span class='op'>)</span>, 
                    Reconstruction <span class='op'>=</span><span class='fl'>0</span>, BP<span class='op'>=</span><span class='fl'>0</span>, Hapfile<span class='op'>=</span><span class='cn'>NULL</span>, Mapfile<span class='op'>=</span><span class='cn'>NULL</span>, seed<span class='op'>=</span> <span class='fl'>0</span>, 
                    outDir <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/getwd.html'>getwd</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gen</th>
      <td><p>An object of class GLgen obtained with gen.genealogy, gen.lineages or gen.branching. Required.</p></td>
    </tr>
    <tr>
      <th>pro</th>
      <td><p>Vector of proband id numbers to be included. Required.</p></td>
    </tr>
    <tr>
      <th>ancestors</th>
      <td><p>Vector of ancestors id numbers to be included</p></td>
    </tr>
    <tr>
      <th>simulNo</th>
      <td><p>Number of simulations to run. Results from extra simulations will be included in the same output text file</p></td>
    </tr>
    <tr>
      <th>RecombRate</th>
      <td><p>The expected number of recombination events in the simulated region per meiosis. E.g.: if simulating a region of 1Centimorgan, recombination rate should be 0.01
    Vector of 2 elements, the first is recombination rate for male offspring, the second element specifies rate for female offspring</p></td>
    </tr>
    <tr>
      <th>Reconstruction</th>
      <td><p>Indicator specifying whether the simulation results should be converted into sequence data.
    Default is Reconstruction = 0, which is no sequence data Reconstruction. 
    If Reconstruction = 1 then user must provide BP, Hapfile, and Mapfile</p></td>
    </tr>
    <tr>
      <th>BP</th>
      <td><p>The length of the simulated region in basepairs. Only required if Reconstruction = 1</p></td>
    </tr>
    <tr>
      <th>Hapfile</th>
      <td><p>The name of a file in the working directory  containing sequence data for the founder haplotypes. Each line specifies a haplotype, and should be the same format as the example</p></td>
    </tr>
    <tr>
      <th>Mapfile</th>
      <td><p>The name of a file in the working directory which specifies the locations of the SNPs of Hapfile. This file should contain only one number per each line, each line specifies the position (in BP) of the corresponding SNP in the hapfile</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>this will seed the random number gennerator (Mersenne twister) for the simulation. Default is seed = 0 which will choose a random seed. Use the seed for replication purposes</p></td>
    </tr>
    <tr>
      <th>outDir</th>
      <td><p>The directory to write output files to. Default is the current working directory.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A matrix with 3 columns: simulNo, NumMeioses, NumRecomb
Will generate output files "Proband_Haplotypes.txt" and "All_nodes_haplotypes.txt" in the specified output directory. If Reconstruction = 1 will also generate "reconstructed_haplotypes.txt"</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='gen.genealogy.html'>gen.genealogy</a></code>
<code><a href='gen.simuSample.html'>gen.simuSample</a></code>
<code><a href='gen.simuSet.html'>gen.simuSet</a></code>
<code><a href='gen.simuSampleFreq.html'>gen.simuSampleFreq</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>#load the genealogy table</span>
<span class='va'>sample_gen_file</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>,<span class='st'>"sample_gen.csv"</span>,package<span class='op'>=</span><span class='st'>"GENLIB"</span><span class='op'>)</span>
<span class='va'>sample_gen_table</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.csv</a></span><span class='op'>(</span><span class='va'>sample_gen_file</span><span class='op'>)</span>

<span class='co'>#convert the genealogy table into a GENLIB object</span>
<span class='va'>sample_gen</span><span class='op'>&lt;-</span><span class='fu'><a href='gen.genealogy.html'>gen.genealogy</a></span><span class='op'>(</span><span class='va'>sample_gen_table</span>, autoComplete <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> 

<span class='co'>#visualize the genealogy</span>
<span class='fu'><a href='gen.graph.html'>gen.graph</a></span><span class='op'>(</span><span class='va'>sample_gen</span><span class='op'>)</span>
</div><div class='img'><img src='gen.simuHaplo-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='va'>probands</span> <span class='op'>&lt;-</span> <span class='fu'><a href='gen.pro.html'>gen.pro</a></span><span class='op'>(</span><span class='va'>sample_gen</span><span class='op'>)</span>
<span class='va'>founders</span> <span class='op'>&lt;-</span> <span class='fu'><a href='gen.founder.html'>gen.founder</a></span><span class='op'>(</span><span class='va'>sample_gen</span><span class='op'>)</span>

<span class='co'># Simulate the gene dropping from founders to probamds</span>
<span class='co'># Generate haplotypes of probands, with segments labelled corresponding to the founder</span>
<span class='co'># If "Reconstruction = 0"  as the default so no sequence data is required, </span>
<span class='co'>#  and the generated haplotypes will be in terms relative position and founder ID</span>

<span class='fu'>gen.simuHaplo</span><span class='op'>(</span><span class='va'>sample_gen</span>,pro <span class='op'>=</span> <span class='va'>probands</span>, ancestors<span class='op'>=</span> <span class='va'>founders</span>, simulNo<span class='op'>=</span><span class='fl'>1</span>, 
                RecombRate<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.01</span>,<span class='fl'>0.02</span><span class='op'>)</span>, seed<span class='op'>=</span><span class='fl'>0</span>, outDir<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>seed: 1377303848</span></div><div class='output co'>#&gt; <span class='message'>output files: C:\Users\Mohan\AppData\Local\Temp\Rtmp0aMWpv/All_nodes_haplotypes.txt </span>
#&gt; <span class='message'>C:\Users\Mohan\AppData\Local\Temp\Rtmp0aMWpv/Proband_Haplotypes.txt </span></div><div class='output co'>#&gt;      simulNo numRecomb numMeioses
#&gt; [1,]       1         0         16</div><div class='input'>
<span class='co'># The output files will be written to the working directory. </span>
<span class='co'># If you do not wish to give access to this directory, </span>
<span class='co'># then change the working directory with the command setwd() before running the function</span>

<span class='co'>#We can also re-run the function with the Reconstruction = 1 option to use sequence data</span>
<span class='co'># this will generate a file of simulated proband haplotypes with sequence data</span>
<span class='co'># But we must provide the function with a Hapfile and Mapfile.</span>

<span class='co'>#Hapfile contains the sequence (SNP) data for each founder haplotype. </span>
<span class='co'># Each line should consist of only 2 elements:</span>
<span class='co'># the founder haplotype ID #, and the sequence, seperated by a single whitespace</span>
<span class='co'># the sequence data must be a string and use any characters, </span>
<span class='co'># but must be the same length between all founders.</span>
<span class='co'># Each position in the string corresponds to a BP position in the haplotype,</span>
<span class='co'># so any missing data should be encoded, using a missing data character and not skipped</span>
<span class='co'># Example hapfile:</span>

<span class='va'>Hapfile</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>,<span class='st'>"sequences.txt"</span>,package<span class='op'>=</span><span class='st'>"GENLIB"</span><span class='op'>)</span>

<span class='co'>#Mapfile contains the base pair positions of all the SNP's in the Hapfile. </span>
<span class='co'>#Each line describes the position of the corresponding character in the sequence</span>
<span class='co'># there should be no empty lines or whitespaces. </span>
<span class='co'># Example mapfile:</span>

<span class='va'>Mapfile</span><span class='op'>&lt;-</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>,<span class='st'>"SNP_positions.txt"</span>,package<span class='op'>=</span><span class='st'>"GENLIB"</span><span class='op'>)</span>

<span class='fu'>gen.simuHaplo</span><span class='op'>(</span><span class='va'>sample_gen</span>,pro <span class='op'>=</span> <span class='va'>probands</span>, ancestors<span class='op'>=</span> <span class='va'>founders</span>, simulNo<span class='op'>=</span><span class='fl'>2</span>, RecombRate<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.01</span>,<span class='fl'>0.02</span><span class='op'>)</span>, 
                Reconstruction <span class='op'>=</span> <span class='fl'>1</span>, BP<span class='op'>=</span> <span class='fl'>1000000</span>, Hapfile<span class='op'>=</span> <span class='va'>Hapfile</span>, Mapfile <span class='op'>=</span> <span class='va'>Mapfile</span>, seed<span class='op'>=</span><span class='fl'>777</span>,
                outDir<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>seed: 777</span></div><div class='output co'>#&gt; <span class='message'>output files: C:\Users\Mohan\AppData\Local\Temp\Rtmp0aMWpv/All_nodes_haplotypes.txt </span>
#&gt; <span class='message'>C:\Users\Mohan\AppData\Local\Temp\Rtmp0aMWpv/Proband_Haplotypes.txt </span>
#&gt; <span class='message'>C:\Users\Mohan\AppData\Local\Temp\Rtmp0aMWpv/reconstructed_haplotypes.txt</span></div><div class='output co'>#&gt;      simulNo numRecomb numMeioses
#&gt; [1,]       1         0         16
#&gt; [2,]       2         0         16</div><div class='input'>
<span class='co'>#Since we used the Reconstruction=1 option, we will now get an additional output file </span>
<span class='co'># with sequence data for the proband haplotypes.</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Louis Houde, Jean-Francois Lefebvre, Valery Roy-Lagace, Sebastien Lemieux, Marie-Helene Roy-Gagnon.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


